File: src/main/java/org/micoli/php/ui/popup/NavigableListPopup.java
package org.micoli.php.ui.popup;

import com.intellij.openapi.ui.popup.PopupChooserBuilder;
import com.intellij.ui.awt.RelativePoint;
import com.intellij.ui.components.JBList;
import java.awt.event.MouseEvent;
import java.util.List;
import javax.swing.*;

public class NavigableListPopup {

    public static void showNavigablePopup(MouseEvent mouseEvent, List<NavigableListPopupItem> elements) {

        JBList<NavigableListPopupItem> list = new JBList<>(elements);
        list.setCellRenderer(new NavigableListCellRenderer());
        list.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);

        new PopupChooserBuilder<>(list)
                .setItemChosenCallback(() -> {
                    NavigableListPopupItem selected = list.getSelectedValue();
                    if (selected != null && selected.canNavigate()) {
                        selected.navigate(true);
                    }
                })
                .createPopup()
                .show(new RelativePoint(mouseEvent));
    }
}

File: src/main/kotlin/org/micoli/php/ui/popup/NavigableListPopup.kt
package org.micoli.php.ui.popup

import com.intellij.openapi.ui.popup.PopupChooserBuilder
import com.intellij.ui.awt.RelativePoint
import com.intellij.ui.components.JBList
import java.awt.event.MouseEvent
import javax.swing.ListSelectionModel

object NavigableListPopup {
    @JvmStatic
    fun showNavigablePopup(mouseEvent: MouseEvent, elements: MutableList<NavigableListPopupItem?>) {
        val list = JBList<NavigableListPopupItem?>(elements)
        list.cellRenderer = NavigableListCellRenderer()
        list.selectionMode = ListSelectionModel.SINGLE_SELECTION

        PopupChooserBuilder<NavigableListPopupItem?>(list)
          .setItemChosenCallback(
            Runnable {
                val selected = list.getSelectedValue()
                if (selected != null && selected.canNavigate()) {
                    selected.navigate(true)
                }
            }
          )
          .createPopup()
          .show(RelativePoint(mouseEvent))
    }
}
